core/templates/dev/head/domain/learner_dashboard/LearnerDashboardIconsDirective.js

```diff
@@ -59,6 +59,20 @@ oppia.directive('learnerDashboardIcons', [
                   response.data.learner_dashboard_activity_ids));
             }
           );
+          var learnerPlaylistUrl = (
+            UrlInterpolationService.interpolateUrl(
+              '/learnerplaylistactivityhandler/<activityType>/<activityId>', {
+                activityType: $scope.getActivityType(),
+                activityId: $scope.getActivityId()
+              }));
+          $http.post(learnerPlaylistUrl, {})
```

Comment:
I don't think this directive should be directly doing $http requests. All contact with $http should be done via BackendApiService.

Also is there some way you can get the data to the frontend in a single initial batch call (perhaps when you pick up the activities in the first place)? Otherwise this is going to result in a whole bunch of simultaneous $http calls to the backend if this directive is being replicated multiple times when there are lots of explorations.

Reply:
I understood the problem here @seanlip I'll shift the code to BackendApiService.

Also, the clock icon is displayed on hovering on a particular activity, that means API call is done for a particular activity. And as the requests are done and compared per activity, I don't think we can make bulk requests.

Reply:
But why can't you get the data at the outset when all the activities are fetched?

Reply:
Okay, I'll try fixing it.

Reply:
Hi @WickedBrat -- I can't look at this in detail now, sorry; this week is a bit of a crazy one for me. But it's fine to drop this for now and take up another issue if the technical design is too tricky.

Reply:
Hi @WickedBrat, maybe you can add `canAddToLearnerPlaylist(activityId, activityType)` in `LearnerPlaylistService` and then inject(add dependency) to the `LearnerDashboardIconsDirective`'s controller?